packages/firestore/src/core/transaction_runner.ts

```diff
@@ -0,0 +1,113 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Deferred } from './../util/promise';
+import { TimerId, AsyncQueue } from './../util/async_queue';
+import { ExponentialBackoff } from './../remote/backoff';
+import { Transaction } from './transaction';
+import { RemoteStore } from '../remote/remote_store';
+import { isNullOrUndefined } from '../util/types';
+import { isPermanentError } from '../remote/rpc_error';
+import { FirestoreError } from '../util/error';
+
+const RETRY_COUNT = 5;
+/**
+ * TransactionRunner encapsulates the logic needed to run and retry transactions
+ * with backoff.
+ */
+export class TransactionRunner<T> {
+  private retries = RETRY_COUNT;
+  private backoff: ExponentialBackoff;
+
+  constructor(
+    private readonly asyncQueue: AsyncQueue,
+    private readonly remoteStore: RemoteStore,
+    private readonly updateFunction: (transaction: Transaction) => Promise<T>,
+    private readonly deferred: Deferred<T>
+  ) {
+    this.backoff = new ExponentialBackoff(
+      this.asyncQueue,
+      TimerId.RetryTransaction
+    );
+  }
+  /** Runs the transaction and returns a Promise containing the result. */
+  async run(): Promise<T> {
+    this.runWithBackOff();
+    return this.deferred.promise;
+  }
+
+  private runWithBackOff(): void {
+    this.backoff.backoffAndRun(async () => {
+      const transaction = this.remoteStore.createTransaction();
+      try {
+        const userPromise = this.updateFunction(transaction);
+        if (
+          isNullOrUndefined(userPromise) ||
+          !userPromise.catch ||
+          !userPromise.then
+        ) {
+          this.deferred.reject(
+            Error('Transaction callback must return a Promise')
+          );
+          return;
+        }
+        userPromise
+          .then(result => {
+            return this.asyncQueue.enqueue(() => {
```

Comment:
I think this should probably be `enqueueAndForget()` since we're not returning the result to anybody (and we don't want errors from here to be handled by the `catch(error => ...)` clause on line 81 below...